File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.cpp
// Last updated: 8/5/2025, 2:54:04 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
       return depth_first_search(root, std::numeric_limits<int>::min());
    }

private:
    int depth_first_search(TreeNode* root, int curr_max) {
        if (!root) {
            return 0;
        }

        int count = 0;
        if (curr_max <= root->val) {
            // Node is good.
            count ++;
            curr_max = root->val;
        }

        count += depth_first_search(root->left, curr_max);
        count += depth_first_search(root->right, curr_max);

        return count;
    }
};

File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.py
# Last updated: 8/5/2025, 2:54:01 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

## DFS Iterative
class Solution:
    def goodNodes(self, root: TreeNode) -> int:

        def dfs(root, prev_max):
            if not root:
                return 0
            
            if root.val >= prev_max:
                return 1 + dfs(root.left, root.val) + dfs(root.right, root.val)
            else:
                return dfs(root.left, prev_max) + dfs(root.right, prev_max)


        return dfs(root, root.val)














        # stack = [(root, float("-inf"))]
        # num_good_nodes = 0
        
        # while stack:
        #     node, max_so_far = stack.pop()
        #     if max_so_far <= node.val:
        #         num_good_nodes += 1
        #     if node.right:
        #         stack.append((node.right, max(node.val, max_so_far)))
        #     if node.left:
        #         stack.append((node.left, max(node.val, max_so_far)))
        
        # return num_good_nodes
        
        # if not root:
        #     return 0
        
        # good_nodes = 0
        # stack = []
        # stack.append((root, root.val))

        # while stack:
        #     curr, maxVal = stack.pop()
        #     if curr.val >= maxVal:
        #         good_nodes += 1
        #         maxVal = curr.val
        #     if curr.right:
        #         stack.append((curr.right, maxVal))
        #     if curr.left:
        #         stack.append((curr.left, maxVal))

        # return good_nodes